### PR TITLE
r/aws_dynamo_contributor_insights: Add new resource

### LINKED ISSUE #13933: Feature Request: DynamoDB CloudWatch Contributor Insights
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

CloudWatch Contributor Insights was made [Generally Available](https://aws.amazon.com/blogs/aws/cloudwatch-contributor-insights-for-dynamodb-now-generally-available/) in April 2020:

_Amazon CloudWatch Contributor Insights analyzes log data and creates time-series visualizations to provide a view of top contributors influencing system performance. You do this by creating Contributor Insights rules to evaluate CloudWatch Logs (including logs from AWS services) and any custom logs sent by your service or on-premises servers._

### New or Affected Resource(s)

* aws_dynamodb_table

### Potential Terraform Configuration

```hcl
resource "aws_dynamodb_table" "enable_via_table" {
  name           = "TerraformTestTable"
  enable_contribution_insights = true
}

resource "aws_dynamodb_table" "enable_via_index" {
  name           = "TerraformTestTable"
  global_secondary_index {
    name            = "InitialTestTableGSI"
    hash_key        = "TestTableHashKey"
    range_key       = "TestGSIRangeKey"
    write_capacity  = 1
    read_capacity   = 1
    projection_type = "KEYS_ONLY"
    enable_contribution_insights = true
  }
}
```

### References

* [AWS API Documentation](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_UpdateContributorInsights.html)
* [AWS Go SDK Documentation](https://docs.aws.amazon.com/sdk-for-go/api/service/dynamodb/#DynamoDB.UpdateContributorInsights)

### PR DESCRIPTION
**Pull request by johnsonaj**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #13933

### Description

Creates new resource for `aws_dynamodb_contributor_insights`

```hcl
resource "aws_dynamodb_contributor_insights" "test" {
  table_name = "TestTableName"
}
```
Output from acceptance testing:

```
$ make testacc TESTS=TestAccContributorInsights_ PKG=dynamodb

=== RUN   TestAccContributorInsights_basic
=== PAUSE TestAccContributorInsights_basic
=== RUN   TestAccContributorInsights_disappears
=== PAUSE TestAccContributorInsights_disappears
=== CONT  TestAccContributorInsights_basic
=== CONT  TestAccContributorInsights_disappears
--- PASS: TestAccContributorInsights_disappears (32.56s)
--- PASS: TestAccContributorInsights_basic (56.66s)
PASS
...
```

### CODE REVIEW

File: internal/service/dynamodb/contributor_insights.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Optional: true,
+			},
+			"table_name": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+			"status": {
```
Comment (ewbankkit) [inline internal/service/dynamodb/contributor_insights.go:44]: Is the `status` useful for the practitioner? We now tend to not include attributes like `status` or `updated_date` in the resource as they often change outside the control of the resource's Terraform configuration. They are often included in any parallel data source.

File: internal/service/dynamodb/contributor_insights.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceContributorInsightsCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*conns.AWSClient).DynamoDBConn
+
+	in := &dynamodb.UpdateContributorInsightsInput{
```
Comment (ewbankkit) [inline internal/service/dynamodb/contributor_insights.go:55]: The current convention seems to be to name any API input variable `input`.

File: internal/service/dynamodb/contributor_insights.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+	var indexName string
+	if v, ok := d.GetOk("index_name"); ok {
+		in.IndexName = aws.String(v.(string))
+		indexName = v.(string)
+	}
+
+	out, err := conn.UpdateContributorInsightsWithContext(ctx, in)
+	if err != nil {
+		return diag.Errorf("creating dynamodb ContributorInsights for table (%s): %s", d.Get("table_name").(string), err)
```
Comment (ewbankkit) [inline internal/service/dynamodb/contributor_insights.go:71]: In other DynamoDB resources, e.g. https://github.com/hashicorp/terraform-provider-aws/blob/690cee18d1d3aefce2270cc88cf21e05b6ae6a06/internal/service/dynamodb/table.go#L547 we are using `DynamoDB` in any error messages.

```suggestion
		return diag.Errorf("creating DynamoDB ContributorInsights for table (%s): %s", d.Get("table_name").(string), err)
```

File: internal/service/dynamodb/contributor_insights.go (hunk 2 above)
Comment (ewbankkit) [reply internal/service/dynamodb/contributor_insights.go:55]: Same in other CRUD handler methods.

File: internal/service/dynamodb/contributor_insights.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+		in.TableName = aws.String(v.(string))
+	}
+
+	var indexName string
+	if v, ok := d.GetOk("index_name"); ok {
+		in.IndexName = aws.String(v.(string))
+		indexName = v.(string)
+	}
+
+	out, err := conn.UpdateContributorInsightsWithContext(ctx, in)
```
[the remark this answers is not in the document]
Comment (ewbankkit) [reply internal/service/dynamodb/contributor_insights.go:69]: Same in other CRUD handler methods.

File: internal/service/dynamodb/contributor_insights.go (hunk 3 above)
Comment (ewbankkit) [reply internal/service/dynamodb/contributor_insights.go:71]: Same in other CRUD handler methods.

File: internal/service/dynamodb/contributor_insights_test.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	sdkacctest "github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+	"github.com/hashicorp/terraform-provider-aws/internal/acctest"
+	"github.com/hashicorp/terraform-provider-aws/internal/conns"
+	tfdynamodb "github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb"
+)
+
+func TestAccContributorInsights_basic(t *testing.T) {
```
Comment (ewbankkit) [inline internal/service/dynamodb/contributor_insights_test.go:20]: Add a `_disappears` test also.

**Comment by ewbankkit** [comment]
##### AWS Commercial

```console
% make testacc TESTS=TestAccContributorInsights_ PKG=dynamodb
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/dynamodb/... -v -count 1 -parallel 20 -run='TestAccContributorInsights_'  -timeout 180m
=== RUN   TestAccContributorInsights_basic
=== PAUSE TestAccContributorInsights_basic
=== CONT  TestAccContributorInsights_basic
--- PASS: TestAccContributorInsights_basic (86.40s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb	89.966s
```

##### AWS GovCloud

```console
% make testacc TESTS=TestAccContributorInsights_ PKG=dynamodb
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/dynamodb/... -v -count 1 -parallel 20 -run='TestAccContributorInsights_'  -timeout 180m
=== RUN   TestAccContributorInsights_basic
=== PAUSE TestAccContributorInsights_basic
=== CONT  TestAccContributorInsights_basic
--- PASS: TestAccContributorInsights_basic (77.25s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb	80.791s
```

File: internal/service/dynamodb/contributor_insights.go (hunk 1 above)
Comment (johnsonaj) [reply internal/service/dynamodb/contributor_insights.go]: Removed `status`

File: internal/service/dynamodb/contributor_insights.go (hunk 2 above)
Comment (johnsonaj) [reply internal/service/dynamodb/contributor_insights.go]: updated variable names

File: internal/service/dynamodb/contributor_insights.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (johnsonaj) [reply internal/service/dynamodb/contributor_insights.go]: updated

File: internal/service/dynamodb/contributor_insights.go (hunk 3 above)
Comment (johnsonaj) [reply internal/service/dynamodb/contributor_insights.go]: updated to follow convention of other errors

File: internal/service/dynamodb/find.go [hunk 6]
```diff
@@ -119,3 +119,20 @@ func FindDynamoDBTTLRDescriptionByTableName(conn *dynamodb.DynamoDB, tableName s
...
+	}
+
+	if indexName != "" {
+		input.IndexName = aws.String(indexName)
+	}
+
+	output, err := conn.DescribeContributorInsightsWithContext(ctx, input)
+	if err != nil {
+		return nil, err
+	}
```
Comment (ewbankkit) [inline internal/service/dynamodb/find.go:135]: Treat disabled as not found.

```suggestion
	output, err := conn.DescribeContributorInsightsWithContext(ctx, input)

        if tfawserr.ErrCodeEquals(err, dynamodb.ErrCodeResourceNotFoundException) {
		return nil, &resource.NotFoundError{
			LastError:   err,
			LastRequest: input,
		}
        }
        
	if err != nil {
		return nil, err
	}
	
	if status := aws.StringValue(output. ContributorInsightsStatus); status == dynamodb.ContributorInsightsStatusDisabled {
		return nil, &resource.NotFoundError{
			Message:     state,
			LastRequest: input,
		}
	}
	
	if output == nil {
		return nil, tfresource.NewEmptyResultError(input)
	}
```

File: internal/service/dynamodb/status.go [hunk 7]
```diff
@@ -184,3 +184,23 @@ func statusDynamoDBTableSES(conn *dynamodb.DynamoDB, tableName string) resource.
...
 	}
 }
+
+func statusContributorInsights(ctx context.Context, conn *dynamodb.DynamoDB, tableName, indexName string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		insight, err := FindContributorInsights(ctx, conn, tableName, indexName)
+
+		if tfawserr.ErrCodeEquals(err, dynamodb.ErrCodeResourceNotFoundException) {
+			return nil, "", nil
+		}
```
Comment (ewbankkit) [inline internal/service/dynamodb/status.go:194]:
```suggestion
		if tfresource.NotFound(err) {
			return nil, "", nil
		}
```

File: internal/service/dynamodb/wait.go [hunk 8]
```diff
@@ -258,3 +258,29 @@ func waitDynamoDBSSEUpdated(conn *dynamodb.DynamoDB, tableName string) (*dynamod
...
+
+	_, err := stateConf.WaitForStateContext(ctx)
+
+	return err
+}
+
+func waitContributorInsightsDeleted(ctx context.Context, conn *dynamodb.DynamoDB, tableName, indexName string, timeout time.Duration) error {
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{dynamodb.ContributorInsightsStatusDisabling},
+		Target:  []string{dynamodb.ContributorInsightsStatusDisabled},
```
Comment (ewbankkit) [inline internal/service/dynamodb/wait.go:278]: Target is empty if resource is not found (see suggested change in `status.go`).

```suggestion
		Target:  []string{},
```

File: internal/service/dynamodb/contributor_insights_test.go [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+		}
+
+		log.Printf("[DEBUG] Checking if DynamoDB Contributor Insights %s exists", rs.Primary.ID)
+
+		tableName, indexName, err := tfdynamodb.DecodeContributorInsightsID(rs.Primary.ID)
+		if err != nil {
+			return err
+		}
+
+		in := &dynamodb.DescribeContributorInsightsInput{
```
Comment (ewbankkit) [inline internal/service/dynamodb/contributor_insights_test.go:154]: Can use `FindContributorInsights` here instead of duplicating the error check logic.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTS=TestAccContributorInsights_ PKG=dynamodb
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/dynamodb/... -v -count 1 -parallel 20 -run='TestAccContributorInsights_'  -timeout 180m
=== RUN   TestAccContributorInsights_basic
=== PAUSE TestAccContributorInsights_basic
=== RUN   TestAccContributorInsights_disappears
=== PAUSE TestAccContributorInsights_disappears
=== CONT  TestAccContributorInsights_basic
=== CONT  TestAccContributorInsights_disappears
--- PASS: TestAccContributorInsights_disappears (36.85s)
--- PASS: TestAccContributorInsights_basic (52.68s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb	56.407s
```

**Pull request merged**